node/silkworm/downloader/internals/measures.hpp

```diff
@@ -0,0 +1,47 @@
+/*
```

Comment:
Please let's try not to create duplicates.
There's already class `stopwatch` in `node/silkworm/common` which does the task.
Eventually integrate that one instead of creating a new one

Reply:
Ok, I had not seen it, I will use stopwatch in the next commit.

However I think, that we need to create a different class, maybe from stopwatch, that can be left in the code but turned off on production. My class MeasureScopeExecTime can be turned off with an ifdef directive, and is not invasive (only 1 line in the scope that you need to measure). stopwatch is more invasive (you need to do 4 calls to do a measure) and is difficult to turn off it because the log code is not in the stopwatch class but is the applicative class.

Reply:
We can turn off the whole silkworm code with one define. That's not the point.
Imho we should not create and recreate variants of the very same scoped tools only for the purpose to have a whole functionality callable with one line.

Reply:
In general I agree with Andrea, we shouldn't create different flavours of the same functionality. But also perhaps we can simplify `StopWatch` a little bit? I also wanted its API to be somewhat simpler when I initially looked at it. IMHO we can do that a later PR; just let's create a task to remove `MeasureScopeExecTime` in favour of potentially simplified `StopWatch`.

Reply:
C'mon guys ... what is "overly complicated" in StopWatch ?
Instantiate a class and call start ?

Reply:
As I said, I had not seen it, and I have just pushed a commit to delete it

Reply:
About StopWatch. Consider the simplest case when you want to measure a single time interval (and do we even need to measure several intervals with the same watch?) The code you need right now  is something like
```
    StopWatch sw;
    sw.start();
    <CODE TO MEASURE>
    const auto [_, duration] = sw.lap();
    std::cout << "Done in " << StopWatch::format(duration) << std::endl;
```
It's fine, but personally I'd prefer something simpler like
```
    StopWatch sw; // starts automatically on construction
    <CODE TO MEASURE>
    sw.stop();
    std::cout << "Done in " << sw << std::endl;
```
But it's a matter of taste, of course. Creating (and especially agreeing on :) good APIs is difficult.

Reply:
> C'mon guys ... what is "overly complicated" in StopWatch ? Instantiate a class and call start ?

Nothing. I think stopwatch is ok if someone has to do measurements with laps.

But if I need to measure only the time consumed by a scope I prefer to write:

```
{
   MeasureScopeExecTime measure_curr_scope("bla bla");
   ...
}
```
to

```
{
    StopWatch measure_curr_scope;                  
    auto start_time = measure_curr_scope.start();  

    ...

    auto [end_time, _] = measure_curr_scope.lap();  // only for test

    log::Trace() << measure_curr_scope.format(end_time - start_time);
}
```

In my opinion the method gets dirty. Also if we turn off StopWatch implementation with a #ifdef than also the log in the method gets dirty.

Reply:
Before I definitely raise my hands and surrender focusing on something else:

> (and do we even need to measure several intervals with the same watch?)

Yes. Just think for example to a sync cycle and within that the timings of single stages

StopWatch is flexible enough to accomodate any use case ...

> Also if we turn off StopWatch implementation with a #ifdef than also the log in the method gets dirty.

Honestly this statement does not make any sense to me. Why should we disable an ubiquitous class ?